docs/src/literate-tutorials/linear_elasticity.jl

```diff
@@ -1,3 +1,354 @@
+# \begin{align*}
+# \boldsymbol{\varepsilon}^\mathrm{vol} &= \frac{\mathrm{tr}(\boldsymbol{\varepsilon})}{3}\boldsymbol{I}, \quad
+# \boldsymbol{\varepsilon}^\mathrm{dev} &= \boldsymbol{\varepsilon} - \boldsymbol{\varepsilon}^\mathrm{vol}
+# \end{align*}
+# ```
+#
+# Starting from Young's modulus, $E$, and Poisson's ratio, $\nu$, the shear and bulk modulus are
+# ```math
+# G = \frac{E}{2(1 + \nu)}, \quad K = \frac{E}{3(1 - 2\nu)}
+# ```
+# Finally, the stiffness tensor can be calculated as the derivative of stress wrt. strain.
+# Due to the linearity, we can calculate this at any point, and choose zero strain.
+Emod = 200e3 # Young's modulus [MPa]
+ν = 0.3      # Poisson's ratio [-]
+
+Gmod = Emod / (2(1 + ν))  # Shear modulus
+Kmod = Emod / (3(1 - 2ν)) # Bulk modulus
+E4 = gradient(ϵ -> 2 * Gmod * dev(ϵ) + 3 * Kmod * vol(ϵ), zero(SymmetricTensor{2,2}));
+
+#md # !!! details "Plane stress instead of plane strain?"
+#md #     In order to change this tutorial to consider plane stress instead of plane strain,
+#md #     the stiffness tensor should be changed to reflect this. The plane stress elasticity
+#md #     stiffness matrix in Voigt notation for engineering shear strains, is given as
+#md #     ```math
+#md #     \underline{\underline{\boldsymbol{E}}} = \frac{E}{1 - \nu^2}\begin{bmatrix}
+#md #     1 & \nu & 0 \\
+#md #     \nu & 1 & 0 \\
+#md #     0 & 0 & (1 - \nu)/2
+#md #     \end{bmatrix}
+#md #     ```
+#md #     This matrix can be converted into the 4th order stiffness tensor as
+#md #     ```julia
+#md #     E_voigt = Emod * [1.0 ν 0.0; ν 1.0 0.0; 0.0 0.0 (1-ν)/2] / (1 - ν^2)
+#md #     E4 = fromvoigt(SymmetricTensor{4,2}, E_voigt)
+#md #     ```
+#md #
+# ### Element routine
+# To calculate the global stiffness matrix, $K_{ij}$, the element routine computes the
+# local stiffness matrix `ke` for a single element and assembles it into the global matrix.
+# `ke` is pre-allocated and reused for all elements.
+#
+# Note that the elastic stiffness tensor $\boldsymbol{\mathsf{E}}$ is constant.
+# Thus is needs to be computed and once and can then be used for all integration points.
+function assemble_cell!(ke, cellvalues, ∂σ∂ε)
+    for q_point in 1:getnquadpoints(cellvalues)
+        ## Get the integration weight for the quadrature point
+        dΩ = getdetJdV(cellvalues, q_point)
+        for i in 1:getnbasefunctions(cellvalues)
+            ## Gradient of the test function
+            ∇Nᵢ = shape_gradient(cellvalues, q_point, i)
+            for j in 1:getnbasefunctions(cellvalues)
+                ## Symmetric gradient of the trial function
+                ∇ˢʸᵐNⱼ = shape_symmetric_gradient(cellvalues, q_point, j)
+                ke[i, j] += (∇Nᵢ ⊡ ∂σ∂ε ⊡ ∇ˢʸᵐNⱼ) * dΩ
+            end
+        end
+    end
+    return ke
+end
+#md nothing #hide
+#
+#md # !!! details "Performance tips"
+#md #     1. In this tutorial, `∂σ∂ε`, is minor symmetric, and we could have used
+#md #        the symmetric part of test function gradient above.
```

Comment:
Why is this a performance tip?

Reply:
Because I thought you'd save some computational time, but seems negligble and only detectable in 3d...
```julia
E4 = rand(SymmetricTensor{4,3});
xs = rand(SymmetricTensor{2,3});
x = rand(Tensor{2,3});
@btime dcontract($E4, $xs); # 3.200 ns (0 allocations: 0 bytes)
@btime dcontract($E4, $x);  # 3.500 ns (0 allocations: 0 bytes)
```